dask/dataframe/core.py

```diff
@@ -2330,6 +2335,236 @@ def quantile(df, q):
+    approximate percentiles for generic objects in an out-of-core fashion is
+    also hard.  Fortunately, partition boundaries don't need to be perfect
+    in order for partitioning to be effective, so we strive for a "good enough"
+    method that can scale to many partitions and is reasonably well-behaved for
+    a wide variety of scenarios.
+
+    Two similar approaches come to mind: (1) take a subsample of every
+    partition, then find the best new partitions for the combined subsamples;
+    and (2) calculate equally-spaced percentiles on every partition (a
+    relatively cheap operation), then merge the results.  We do both, but
+    instead of random samples, we use random percentiles.
+
+    If the number of partitions isn't changing, then the ratio of fixed
+    percentiles to random percentiles is 2 to 1.  If repartitioning goes from
+    a very high number of partitions to a very low number of partitions, then
+    we use more random percentiles, because a stochastic approach will be more
+    stable to potential correlations in the data that may cause a few equally-
+    spaced partitions to under-sample the data.
+
+    The more partitions there are, then the more total percentiles will get
+    calculated across all partitions.  Squaring the number of partitions
+    approximately doubles the number of total percentiles calculated, so
+    num_total_percentiles ~ sqrt(num_partitions).  We assume each partition
+    is approximately the same length.  This should provide adequate resolution
+    and allow the number of partitions to scale.
+
+    For numeric data, one could instead use T-Digest for floats and Q-Digest
+    for ints to calculate approximate percentiles.  Our current method works
+    for any dtype.
+    """
+    # *waves hands*
+    random_percentage = 1 / (1 + (4.0 * num_new / num_old)**0.5)
+    num_percentiles = upsample * num_new * (num_old + 22)**0.55 / num_old
+    num_fixed = int(num_percentiles * (1 - random_percentage)) + 2
+    num_random = int(num_percentiles * random_percentage) + 2
+
+    if num_fixed + num_random + 5 >= chunk_length:
+        return np.linspace(0, 100, chunk_length + 1)
+
+    if not isinstance(random_state, np.random.RandomState):
+        random_state = np.random.RandomState(random_state)
+
+    q_fixed = np.linspace(0, 100, num_fixed)
+    q_random = random_state.rand(num_random) * 100
+    q_edges = [0.5 / num_fixed, 1 - 0.5 / num_fixed]
+    qs = np.concatenate([q_fixed, q_random, q_edges])
+    qs.sort()
+    return qs
+
+
+def tree_width(N, to_binary=False):
+    """Generate tree width suitable for ``merge_sorted`` given N inputs
+
+    In theory, this is designed so all tasks are of comparable effort.
+    """
+    if N < 32:
+        group_size = 2
+    else:
+        group_size = int(math.log(N))
+    num_groups = N // group_size
+    if to_binary or num_groups < 16:
+        return 2**int(math.log(float(N) / group_size, 2))
+    else:
+        return num_groups
```

Comment:
:)

At one point, `sympy` included `LambertW` in the solution.  It's just a happy coincidence that `group_size = int(math.log(N))`results in a remarkably close approximation.  The goal was that tasks for `merge_sorted`, which scales as `N * lg(k)` would be 1-2x the effort of the root node.